lib/Parse/ParseExpr.cpp

```diff
@@ -2744,6 +2744,79 @@ void Parser::ParseBlockId(SourceLocation CaretLoc) {
   Actions.ActOnBlockArguments(CaretLoc, DeclaratorInfo, getCurScope());
 }
 
+IdentifierInfo *Parser::getBoundsExpressionStart() {
+  if (Tok.getKind() == tok::identifier) {
+    IdentifierInfo *Ident = Tok.getIdentifierInfo();
+    if (Ident == Ident_bounds || Ident == Ident_byte_count || Ident == Ident_count) {
+      return Ident;
+    }
+  }
+  return nullptr;
+}
+
+ExprResult Parser::ParseBoundsExpression() {
```

Comment:
There's a lot going on in here, and in the future there will be more; some more inline comments might help guide the reader to more quickly parse it.